### PR TITLE
UI Card extension

### PR DESCRIPTION
**Pull request by xus**
Pull Request to discuss the interface of the new UI Card type.
Refactor for UI Card to allow other types.
New Card type ->  repositoryObject.
Needed for -> https://docu.ilias.de/goto_docu_wiki_wpage_4921_1357.html
See examples

### CODE REVIEW

**Comment by xus** [comment]
Before having the proper icons this feature looks like this:
**Small devices**
small

**Desktop and wide resolutions**
We have a lot of space between icons.
wide

If the icons should link somewhere, I will change the icons for Bulky buttons. But then we have to decide how this Bulky button has to be presented.
 - Expand the button to the full column?
 - Hover effect?
 - Any border for the button or row?
 - Bulky button comes with a default grey color. Should be modified to provide a transparent background?.

with_bulky

**Progressmeter issue:**
**Mini version** does not provide the numerical value and has background color.
screen shot 2018-09-24 at 11 34 18

**Standard version** provides the numerical value but it does not scale to low resolutions and also has background color.
screen shot 2018-09-24 at 11 34 31

**Comment by matthiaskunkel** [comment]
I think we need to define a max-width for the tiles. It doesn't make sense to give them as much width as available. That would be not handable for designers, especially when you want to use a photo (which will happen often hopefully).

File: src/UI/Component/Card/Factory.php
Comment (Amstutz) [inline src/UI/Component/Card/Factory.php]: This is a 1:1 copy from the parent class card. this is redundant, please rephrase and only point out, what differntiates this card from other cards (this might be short). see other "standard" examples in other families.

File: src/UI/Component/Card/Factory.php
Comment (Amstutz) [inline src/UI/Component/Card/Factory.php]: see comment above.

File: src/UI/Component/Card/Factory.php
Comment (Amstutz) [inline src/UI/Component/Card/Factory.php]: The complete part is copied from the base factory entry for cards. This is not the idea here.

File: src/UI/Component/Card/Factory.php
Comment (Amstutz) [inline src/UI/Component/Card/Factory.php]: Do not use lower camelCase in descriptions. Use Repository Object instead.

File: src/UI/Component/Card/Factory.php
Comment (Amstutz) [inline src/UI/Component/Card/Factory.php]: This part is again, copied from the parent...

File: src/UI/Component/Card/Factory.php
Comment (Amstutz) [inline src/UI/Component/Card/Factory.php]: Obviously, they are both cards... no need to tell that here.

File: src/UI/Component/Card/Factory.php
Comment (Amstutz) [inline src/UI/Component/Card/Factory.php:50]: Good.

File: src/UI/Component/Card/Factory.php
Comment (Amstutz) [inline src/UI/Component/Card/Factory.php]: Why do you list rules here. See the base templates or any other examples on how and where to place rules.

File: src/UI/Component/Card/Factory.php
Comment (Amstutz) [inline src/UI/Component/Card/Factory.php:52]: again copied from the base class.

File: src/UI/Component/Card/RepositoryObject.php
Comment (Amstutz) [inline src/UI/Component/Card/RepositoryObject.php:47]: are you sure you want to pass an icon here? The card could pick the correct icon by it's own. Will this be linked to a cert (you listed the icon under effect above)? If so just pass the link, if one is set the icon will be displayed. if not, just pass a bool.

**Review by Amstutz — CHANGES_REQUESTED** [review]
@xus thanks for the PR.

In general I think this is heading into the right direction. Adding a new sub type here seems to be correct. Due to the 4 proposed cell I guess this is a good name (e.g. Learning Progress seems very Repository specific).

I did not comment on the implementation, only the interface. This is too far away from a final solution to be accepted today please revise first.

You may discuss concrete question on specific issues however.

**Comment by xus** [comment]
Thank you @Amstutz for your comments, I just copy/pasted all the description from the old Card. I removed the redundant text from the Standard type.

File: src/UI/Component/Card/RepositoryObject.php
Comment (alex40724) [reply src/UI/Component/Card/RepositoryObject.php:47]: We prefer to use only a bool parameter and no link / certificate.

**Comment by alex40724** [comment]
Jour Fixe, 24 SEP 2018 : We highly appreciate this suggestion. We decidec that icons in the snowboard should not be clickable but only the background image below and the title. We will use the outline icons offered by CaT for presenting the object type and the certificate. For the progress meter we will use the mini version plus text (and without background). We will try to solve the responsive problem of the progress meter as well.

**Comment by xus** [comment]
Pending tasks:

- [x] Create specific description/rules for the interfaces. Remove duplicated specifications.
- [x] Use a boolean as a parameter for withCertificatedIcon method and generate the icon in the renderer.
- [x] Css for Icons and Dropdown
- [x] Avoid super big icons in wide resolutions
- [x] Deck of cards adaptation

- [x] Background image clickable.
I made the image clicable for the base card object.
I wrote mutators and getters for the clickable image. I was thinking about create generic methods for actions in titles and images (withAction($url), getAction()) instead of having individual methods but I thought we can keep this open for the code consumer. (e.g. in other projects the links to a course and the background image opens the full image) 
@Amstutz  what do you think?


- [ ] Responsive Progressmeter (https://mantis.ilias.de/view.php?id=23654)

- [ ] Progressmeter with transparent background (https://mantis.ilias.de/view.php?id=23653)

**Comment by klees** [comment]
I'm fine with this extension for now, but we need to clarify the relation between `Item` and `Card`. To me that new `Card` is just an item with a certain rendering. I'll try to outline an according proposal for the UI-frameworks roadmap.

**Comment by Amstutz** [comment]
I removed the JF label, the decision was made in the last JF. 

@xus is this ready for a final review (apart from the resp. progressmeter)?

**Comment by xus** [comment]
Hi @Amstutz @klees PR ready for the final review

**Comment by Amstutz** [comment]
I will wait for the JF decision on: https://github.com/ILIAS-eLearning/ILIAS/pull/1228.

File: src/UI/examples/Card/RepositoryObject/with_object_icon_and_actions.php
Comment (Amstutz) [inline src/UI/examples/Card/RepositoryObject/with_object_icon_and_actions.php]: Why would you use a custom icon here? Note that course is a standard icon. Take it from the standard set please.

File: src/UI/examples/Card/RepositoryObject/with_object_icon_and_certificate.php
Comment (Amstutz) [inline src/UI/examples/Card/RepositoryObject/with_object_icon_and_certificate.php]: again, use the standard set of icons.

File: src/UI/templates/default/Card/card.less
Comment (Amstutz) [inline src/UI/templates/default/Card/card.less]: No colors here, move that to the variables section, see: https://github.com/ILIAS-eLearning/ILIAS/tree/release_5-3/templates#colors

File: src/UI/templates/default/Card/card.less
Comment (Amstutz) [inline src/UI/templates/default/Card/card.less:57]: Where do does magic numbers come from? how are they calculated?

File: src/UI/templates/default/Card/card.less
Comment (Amstutz) [inline src/UI/templates/default/Card/card.less]: No colors here, move that to the variables section, see: https://github.com/ILIAS-eLearning/ILIAS/tree/release_5-3/templates#colors

File: src/UI/templates/default/Card/card.less
Comment (Amstutz) [inline src/UI/templates/default/Card/card.less:69]: Again, where do those numbers come from?

File: src/UI/templates/default/Card/card.less
Comment (Amstutz) [inline src/UI/templates/default/Card/card.less]: No colors here, move that to the variables section, see: https://github.com/ILIAS-eLearning/ILIAS/tree/release_5-3/templates#colors

File: src/UI/templates/default/Icon/icon.less
Comment (Amstutz) [inline src/UI/templates/default/Icon/icon.less:64]: Did you check if that also works with the standard version of the icon?

**Review by Amstutz — CHANGES_REQUESTED** [review]
In Relation to: #1228

Hey @xus: The outlined Icons have been accepted. However, Yvonne needs some time to finish up the final set in collaboration with CaT. Further I do not like to merge my own PR's. So I will wait with that until Richard is back an gives his final review. If everything else is resolved I will merge this without the outlined Icons yet available so you can proceed. Ill adapt this the the outlined icons myself, a soon as they are done. So do not worry about this dependencies.

Since say you need to proceed "ASAP" (#1228) and claim that this is ready for a final review, here we go:

This is not ready for a final review. To start with the essentials: 
The tests fail. This is easily visible on the PR itself.
image
Or just bellow the comment in which you ask for a final review:
image

Further you did not add a single unit test, eventough you propose a new entry. Please consult the following guideline before processing any further:
https://github.com/ILIAS-eLearning/ILIAS/tree/trunk/src/UI#implementing-elements-in-the-framework

Please also consult the less guideline at: https://github.com/ILIAS-eLearning/ILIAS/tree/release_5-3/templates#colors

There are other comments, however in short it is:

- [ ] Use the standard set of Icons. 
- [ ] Revise the less -> introduce variables.
- [ ] Fix existing Tests.
- [ ] Add new Tests to cover the added functionality.

**Comment by xus** [comment]
@Amstutz you are right. I asked for a final review to early, sorry for that.
I will provide the tests as soon as possible.
Also, you are right about the usage of standard icons instead of the custom ones. I'm trying to use them and they are not working properly.
@alex40724 can you please provide feedback to Timon about the changes in the less files?

File: src/UI/templates/default/Card/card.less
Comment (alex40724) [reply src/UI/templates/default/Card/card.less:69]: These are my numbers and for me the best way to give an auto-scaling image a padding that scales together with the image. Is there is a better solution for this problem?

File: src/UI/templates/default/Icon/icon.less
Comment (alex40724) [reply src/UI/templates/default/Icon/icon.less:64]: The standard version does not use img tags. So these rules do not apply. Maybe we currently do not allow standard icons in the repo cards. Not sure how we get the standard version to auto-scale.

File: src/UI/templates/default/Card/card.less
Comment (alex40724) [reply src/UI/templates/default/Card/card.less:57]: What do you mean "how they are calculated"? The whole card is scaling and I try to give some padding that scales with the card.

File: src/UI/examples/Card/RepositoryObject/with_object_icon_and_certificate.php
Comment (alex40724) [reply src/UI/examples/Card/RepositoryObject/with_object_icon_and_certificate.php]: We are implementing this on the base of the custom icons since this is the only way we can assure that we can deliver the feature. We are asked by JF to use the files for the outlined icons, which is perfectly possible using the custom icons. JF did not ask us to introduce a new KS element for outlined icons. Please accept this on the base of custom icons. I am happy if you move this to the new KS outlined icon afterwards.

Comment (Amstutz) [reply src/UI/examples/Card/RepositoryObject/with_object_icon_and_certificate.php]: Hi @alex40724 the JF accepted the interface which requires an Icon which is either a standard or custom. There might be cases, where you will need a custom Icon. However this is surely not one of those. Read the description of the custom Icon:

> ILIAS allows users to upload icons for repository objects. Those, in opposite to the standard icons, need to be constructed with a path.

What would be the purpose of having the standard icon if they are not use for the standard icon set?

`We are asked by JF to use the files for the outlined icons, which is perfectly possible using the custom icons. JF did not ask us to introduce a new KS element for outlined icons. `

Nope the JF did not. I did that (for you). The JF accepted my proposal for those outlined Icons on monday.  Further, note that the outlined Icons are also standard icons, just with an addtional flag set. All you need to do is, make sure, that this works with the standard Icon. If so, the outlined Icons will work as well (since they ARE standard icons and rendered in the exact same manner see: #1228). 

If standard icons mess up the CSS, let me know if you need help. I am sure, we can fix that.

Comment (Amstutz) [reply src/UI/examples/Card/RepositoryObject/with_object_icon_and_certificate.php]: Maybe it's also worth to point out again, that we do not need to wait for the merge of the outlined Icons to merge this. I'll do so, as soon as the pointed out issues are fixed. I expect the outlined Icons to be merged in the middle of next week (when @klees is back for the final review and merge and Yvonne has the completed set ready).

Comment (alex40724) [reply src/UI/examples/Card/RepositoryObject/with_object_icon_and_certificate.php]: Am not sure whether we refer to the same JF decision: "24 SEP 2018: We highly appreciate this suggestion. We decidec that icons in the snowboard should not be clickable but only the background image below and the title. We will use the outline icons offered by CaT for presenting the object type and the certificate. For the progress meter we will use the mini version plus text (and without background). We will try to solve the responsive problem of the progress meter as well."

Does it say we must implement or use a new KS element? I am confused, maybe this is written somewhere else?

Comment (Amstutz) [reply src/UI/examples/Card/RepositoryObject/with_object_icon_and_certificate.php]: on the 24. Sep. we accepted the interface which accepts icons. This monday we accepted outlined icons as variation of the standard icon. using a course icon and not chosing the standard is wrong for the reasons pointed out before. Did you do that to lather switch to the outlined version in the custom version without hassle? 

Please tell me, what the issue is here with using the standard (i guess some css, since we have a background image instead of  as the custom ones do), so we can solve it and move the development on. I already offered to help fixing the CSS problems if such arise.

File: src/UI/templates/default/Card/card.less
Comment (Amstutz) [reply src/UI/templates/default/Card/card.less:57]: > What do you mean "how they are calculated"? 

If it was Trial and Error this is completely fine. I was wondering if they are in any relations to padding variables we already have or are offered by bootstrap. E.g. why 5px and not @padding-large-verticalv (6px) or 2px and not @padding-base-vertical (3px). Maybe you had a sounding reason to make this a px smaller here. If there are relation calculate from the less variables. If you do not believe so, leave as is. You can judge this better here than I do.

For the percentage value, I am not sure if we should do that here. Bootstrap e.g. uses for it's grid columns (which also scale the width) a thing called @grid-gutter-width in px:

```
//** Padding between columns. Gets divided in half for the left and right.
@grid-gutter-width:         30px;
```

They then do something like this when calculating the grid:
>       padding-left:  ceil((@grid-gutter-width / 2));

However, if you take px, you might need 2 variations for this (small and regular) to adapt for screen sizes. This is up to you. If you believe this to be better, leave as is.

File: src/UI/templates/default/Card/card.less
Comment (Amstutz) [reply src/UI/templates/default/Card/card.less:69]: See my comment above, I am not sure about the padding percentages. However, you are more the expert on this less lines for the card, so I leave it up to you to leave this as is.

File: src/UI/templates/default/Icon/icon.less
Comment (Amstutz) [reply src/UI/templates/default/Icon/icon.less:64]: i.O.., I you want me, I can give it a try to make this workable with the standard icon. Just give me a notification if so. Note that there is no way, this will be merged without being able to handle the standard icon.

File: src/UI/examples/Card/RepositoryObject/with_object_icon_and_certificate.php
Comment (alex40724) [reply src/UI/examples/Card/RepositoryObject/with_object_icon_and_certificate.php]: ILIAS repository objects allow to upload icons. We only can use custom icons here. As soon as we allow uploading outlined versions for repository objects (maybe with the next release) we will use custom icons anyway, or am I wrong? This is why we focused on making them auto-scalable. This is easy with images. I did not work out a solution for divs with background images here, since I thought it is acceptable to use the custom icons for now.

I also read the JF decision this way.

If you can provide CSS that autoscales the standard icons in any container we are happy to use this.

File: src/UI/templates/default/Icon/icon.less
Comment (alex40724) [reply src/UI/templates/default/Icon/icon.less:64]: Yes, please try. Quite a disappointing statement though. I focused on custom icons since I thought this would be accepted, since icons are uploadable (even if not the outlined version yet).

File: src/UI/Component/Card/RepositoryObject.php
Comment (alex40724) [reply src/UI/Component/Card/RepositoryObject.php:47]: This is a bool now. Not sure why this is not flagged as "outdated".

**Comment by alex40724** [comment]
All tests work for me now.

File: src/UI/templates/default/Card/card.less
Comment (alex40724) [reply src/UI/templates/default/Card/card.less:57]: I cannot adapt for screen size, since the size of a card does not depend on the screen size only.  The padding needs to scale with its container (the card) imo. If I need to calculate this based on the screen size, it would depend on the the decks less (and breakpoints) suddenly, too (since they define the size of the card "per screen size by defining the cards per row). Please checkout the branch and test it to see how it works. I am happy to see any alternative solution, but I have currently no idea myself.

I also think that the single values in the "padding:5px 4% 2px 4%;" group should not depend on variables that "just meet their values". If someone changes @padding-base-vertical (when used for one of these values) he would not be aware that he is messing up this layout here.

File: src/UI/templates/default/Card/card.less
Comment (alex40724) [reply src/UI/templates/default/Card/card.less:69]: See my comment above.

File: src/UI/templates/default/Card/card.less
Comment (alex40724) [reply src/UI/templates/default/Card/card.less]: This has been moved.

File: src/UI/templates/default/Card/card.less
Comment (alex40724) [reply src/UI/templates/default/Card/card.less]: This has been removed.

File: src/UI/templates/default/Card/card.less
Comment (alex40724) [reply src/UI/templates/default/Card/card.less]: This has been moved.

**Comment by alex40724** [comment]
Less should be ok, too now.

**Comment by alex40724** [comment]
bildschirmfoto 2018-10-22 um 21 40 37

This is how it current presentation.

File: src/UI/examples/Card/RepositoryObject/with_object_icon_and_certificate.php
Comment (Amstutz) [reply src/UI/examples/Card/RepositoryObject/with_object_icon_and_certificate.php]: Hi @alex40724 see my comment bellow.

File: src/UI/templates/default/Icon/icon.less
Comment (Amstutz) [reply src/UI/templates/default/Icon/icon.less:64]: Hi @alex, the following points might help to ease your disappointment about my decision:

1. You need to accept both anyways. In my comment above I wrote.

> Hi @alex40724 the JF accepted the interface which requires an Icon which is either a standard or custom. There might be cases, where you will need a custom Icon.

You or @xus proposed in the PR for the interface to support "Icons" for this, it was not me.  This was fine for me and the JF. This was accepted. If you are proposing to accept Icons, you have to be able to handle them. How else, would you explain somebody, that he may stuff a standard icon in there through the interface, but this completely breaks the css? We could talk about changing the interface, if it was obvious, that clearly only the custom icons make sense here. However this is not the case. Your work on the custom Icon is not lost. You would have to handle them as well, since you accept them on the interface. 

2. You CSS currently completely breaks if no Icon is given. This is possible, again through your own proposal:

> Usage
> Repository Object Cards MAY contain a UI Icon displaying the object type.
> Repository Object Cards MAY contain a UI ProgressMeter displaying the learning progress of the user.
> Repository Object Cards MAY contain a UI Icon displaying a certificate icon if the user finished the task.
> Repository Object Cards MAY contain a UI ProgressMeter OR UI Icon certificate, NOT both.

So you need to address the issue of not having a height set in the il-card-repository-head if not having an custom icon anyways.

3. This is fixable, by:
1) Giving this section a fixed height. However, we would need to make some adaptions for various possible screen AND card sizes. However, this would be solid and also make the rendering of the dropdown a little easier and better (currently it looks funky on various screen/card sizes).

2) Add a completely transparent 32-32 svg which we tell the icon to use as placolder for standard icons. This is no breaking change so no JF decision needed. However, this would not yet solve the "no-icon-at-all" case. For this you would need to add some css fixing the height. If we decide to add the placeholder svg to the standard set, you could use that.

As you see, this is an issue you need to fix regardless of the support of the standard icon question. @alex40724 which option do you prefer? @klees which option do you prefer.

File: src/UI/Implementation/Component/Card/Renderer.php
Comment (Amstutz) [inline src/UI/Implementation/Component/Card/Renderer.php]: This dep. just occured to me, while looking into the "standard icon" issue. Again, here as well: use the standard icon. Note that this is a highly problematic dep. which you can get rid of this way? This icon here is surely not user uploaded content, right?

File: src/UI/templates/default/Icon/icon.less
Comment (klees) [reply src/UI/templates/default/Icon/icon.less:64]: @Amstutz IMO 1. is the way to go.

Comment (alex40724) [reply src/UI/templates/default/Icon/icon.less:64]: My problem is not that I have to fix things, but a) when, b) for what use cases and c) if they are fixable at all. I was not able to work out a proper soloution with a fixed height approach. Did you try this for all deck variants, did it look less funky? It is not so easy if the icons should be equally distributed in the header section. You either have lots of space and tiny icons or bigger icons (I went for the second option, since it looks better to me).

Of course it would be a lot easier if I put all the icons, e.g. right aligned and give them fixed sizes. But this was not the idea of the initial design proposal, right?

I thought it may be acceptable to use the custom icons for now. We do not have another use case for this and could discuss how to handle standard icons later. Also the case that no icon at all is passed is not one that is needed by any feature yet.

I have no JF before beta anymore so I cannt change the rules (and e.g. only accept custom icons for now, or ask the consumer to include at least one element for the header). So I am lost, it just stalls the whole feature.

I very much appreciate if you could try to work out a CSS solution that is acceptable for you.

**Comment by Amstutz** [comment]
@xus and @alex40724 : To push this forward please:
- [x] Remove the toxic ilUtil usage to load the certificate. Use the standard icon instead.
- [x] Fix the examples making use of the standard examples.

If done, ping me, I will merge this and immediatly fix the CSS to work with the standard icons (and with that also outlined icons, see: https://github.com/ILIAS-eLearning/ILIAS/pull/1228).

Note that I regard that as a service to you and that such procedure is an exception. I think the delivered work on the CSS is actually insufficient to be merged (see comments above).

**Comment by alex40724** [comment]
@Amstutz

**Comment by Amstutz** [comment]
I will only do such a thing this one time: https://mantis.ilias.de/view.php?id=23795

**Pull request merged**
